pkg/osvscanner/customgitignore/dir_test.go

```diff
@@ -0,0 +1,309 @@
+	hasMatch = slices.ContainsFunc(patterns, func(p gitignore.Pattern) bool {
+		return p.Match([]string{".", "REPO_EXCLUDE_FILE"}, false) == gitignore.Exclude
+	})
+
+	if !hasMatch {
+		t.Errorf("Expected to find a pattern matching REPO_EXCLUDE_FILE from ./.git/info/exclude ")
+	}
+
+	// expect ./.gitignore to be processed
+	hasMatch = slices.ContainsFunc(patterns, func(p gitignore.Pattern) bool {
+		return p.Match([]string{".", "ROOT_GITIGNORE"}, false) == gitignore.Exclude
+	})
+
+	if !hasMatch {
+		t.Errorf("Expected to find a pattern matching ROOT_GITIGNORE from repository-root .gitignore")
+	}
+
+	// expect ./dir_a/.gitignore to be processed
+	hasMatch = slices.ContainsFunc(patterns, func(p gitignore.Pattern) bool {
+		return p.Match([]string{".", "dir_a", "DIR_A_GITIGNORE"}, false) == gitignore.Exclude
+	})
+
+	if !hasMatch {
+		t.Errorf("Expected to find a pattern matching dir_a/DIR_A_GITIGNORE from ./dir_a/.gitignore")
+	}
+
+	// expect ./parallel_a/.gitignore to be processed
+	hasMatch = slices.ContainsFunc(patterns, func(p gitignore.Pattern) bool {
+		return p.Match([]string{".", "parallel_a", "PARALLEL_A_GITIGNORE"}, false) == gitignore.Exclude
+	})
+
+	if !hasMatch {
+		t.Errorf("Expected to find a pattern matching parallel_a/PARALLEL_A_GITIGNORE from ./dir_a/.gitignore")
+	}
+
+	// expect ./dir_a/dir_b/.gitignore to be skipped over
+	//
+	// I want to test for the lack of a GITIGNORE_B match,
+	// to show that dir_a/dir_b/.gitignore wasn't processed.
+	// Annoyingly the `dir_a/dir_b` pattern from .GITIGNORE_ROOT
+	// prevents this.
+	// Instead I'm doing dubious `reflect` hackery to get
+	// access to the unxported `Pattern.pattern` field
+	for _, pattern := range patterns {
+		// dubious `reflect` hackery means: slice := pattern.pattern
+		fv := reflect.ValueOf(pattern).Elem().FieldByName("pattern")
+
+		// tests if pattern.pattern == []string{"DIR_B_GITIGNORE"}
+		if fv.Len() == 1 && fv.Index(0).String() == "DIR_B_GITIGNORE" {
+			t.Errorf("Expected not to find pattern matching DIR_B_GITIGNORE from ./dir_a/dir_b/.gitignore; " +
+				"dir_b should have been ignored by a rule in repository-root .gitignore")
+		}
+	}
+
+	// expect ./dir_a/dir_b/dir_c/.gitignore to be skipped over
+	// NOTE: see usage of reflect above to see what's going on
+	for _, pattern := range patterns {
+		// dubious `reflect` hackery means: slice := pattern.pattern
+		fv := reflect.ValueOf(pattern).Elem().FieldByName("pattern")
+
+		// tests if pattern.pattern == []string{"DIR_B_GITIGNORE"}
+		if fv.Len() == 1 && fv.Index(0).String() == "DIR_C_GITIGNORE" {
+			t.Errorf("Expected not to find pattern matching DIR_C_GITIGNORE from ./dir_a/dir_b/dir_c/.gitignore; " +
+				"because it's parent dir_b should have been ignored by a rule in repository-root .gitignore")
+		}
+	}
+}
+
+func setupGitRepo(t *testing.T) string {
+	t.Helper()
+
+	// A unique tempdir for local funcs to create a git repo inside
+	gitRepo := filepath.Join(t.TempDir(), "git_repo")
```

Comment:
let's use `testutility.CreateTestDir` (once #832 lands) for this, and you want to return the cleanup function to be `defer`'d in each test

Reply:
Lets do that. We don't need a `defer`-ed cleanup because that's already done for us by `t.Tempdir()` [LINK](https://pkg.go.dev/testing#T.TempDir)

Reply:
[`t.Tempdir` is not suitable due our snapshotting](https://github.com/google/osv-scanner/pull/832#discussion_r1508090735), and I think we should be consistent with what we're using - but #832 switches to using `t.Cleanup` which means you still won't have to do the `defer`

Reply:
Sure. Was there any action you wanted now, besides "wait the future"?

Reply:
Ideally for now I'd pull in the helper like we are elsewhere - that way its consistent pre-#832 and so if this gets merged before that, that PR will get rebased to move this helper and visa versa if #832 get merged first.

Though I expect it'll get landed by Tuesday so probably best to just wait

Reply:
Have added a a TODO for this so it's not lost.